File: Languages/cpp/algorithms-dsa/leetcode/graphs/cheapest_flights_k_stops.cpp
/*
 * LeetCode 787 - Cheapest Flights Within K Stops
 * Topic: Graphs (Shortest Path with Constraints)
 * Difficulty: Medium
 *
 * BFS / Bellman-Ford variant with at most k stops.
 * Relax all edges up to k+1 times (k stops = k+1 edges).
 *
 * Time: O(k * E)
 * Space: O(V)
 */
#include <iostream>
#include <vector>
#include <climits>
#include <cassert>
using namespace std;

int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int k) {
    // Bellman-Ford with k+1 iterations
    vector<int> dist(n, INT_MAX);
    dist[src] = 0;

    for (int i = 0; i <= k; i++) {
        // Copy current distances to avoid using updates from this iteration
        vector<int> prev(dist);
        for (auto& f : flights) {
            int u = f[0], v = f[1], w = f[2];
            if (prev[u] != INT_MAX && prev[u] + w < dist[v]) {
                dist[v] = prev[u] + w;
            }
        }
    }

    return dist[dst] == INT_MAX ? -1 : dist[dst];
}

int main() {
    // Test 1: Basic path with stops limit
    int n1 = 4;
    vector<vector<int>> flights1 = {{0,1,100},{1,2,100},{2,3,100},{0,3,500}};
    assert(findCheapestPrice(n1, flights1, 0, 3, 1) == 500); // Can't use 0->1->2->3 (2 stops)
    assert(findCheapestPrice(n1, flights1, 0, 3, 2) == 300); // 0->1->2->3

    // Test 2: No path
    int n2 = 3;
    vector<vector<int>> flights2 = {{0,1,100},{1,2,100}};
    assert(findCheapestPrice(n2, flights2, 0, 2, 0) == -1);

    // Test 3: Direct flight
    int n3 = 3;
    vector<vector<int>> flights3 = {{0,1,100},{1,2,100},{0,2,500}};
    assert(findCheapestPrice(n3, flights3, 0, 2, 1) == 200);
    assert(findCheapestPrice(n3, flights3, 0, 2, 0) == 500);

    // Test 4: Source equals destination
    int n4 = 2;
    vector<vector<int>> flights4 = {{0,1,100}};
    assert(findCheapestPrice(n4, flights4, 0, 0, 0) == 0);

    cout << "All tests passed!" << endl;
    return 0;
}

File: Languages/python/algorithms-dsa/leetcode/graphs/cheapest_flights_k_stops.py
"""
LeetCode #787 - Cheapest Flights Within K Stops
Topic: Graphs (Shortest Path with Constraints)
Difficulty: Medium

BFS with level limit (Bellman-Ford variant).
Relax all edges up to k+1 times (k stops = k+1 edges).

Time Complexity: O(k * E)
Space Complexity: O(V)
"""


def find_cheapest_price(n: int, flights: list[list[int]], src: int, dst: int, k: int) -> int:
    # Bellman-Ford with at most k+1 relaxation rounds
    dist = [float("inf")] * n
    dist[src] = 0

    for _ in range(k + 1):
        # Use copy to avoid using current iteration's updates
        prev = dist[:]
        for u, v, w in flights:
            if prev[u] != float("inf") and prev[u] + w < dist[v]:
                dist[v] = prev[u] + w

    return dist[dst] if dist[dst] != float("inf") else -1


if __name__ == "__main__":
    # Test 1: Basic path with stops limit
    flights1 = [[0,1,100],[1,2,100],[2,3,100],[0,3,500]]
    assert find_cheapest_price(4, flights1, 0, 3, 1) == 500  # Can't use 0->1->2->3
    assert find_cheapest_price(4, flights1, 0, 3, 2) == 300  # 0->1->2->3

    # Test 2: No path within k stops
    flights2 = [[0,1,100],[1,2,100]]
    assert find_cheapest_price(3, flights2, 0, 2, 0) == -1

    # Test 3: Direct flight vs cheaper path
    flights3 = [[0,1,100],[1,2,100],[0,2,500]]
    assert find_cheapest_price(3, flights3, 0, 2, 1) == 200
    assert find_cheapest_price(3, flights3, 0, 2, 0) == 500

    # Test 4: Source equals destination
    flights4 = [[0,1,100]]
    assert find_cheapest_price(2, flights4, 0, 0, 0) == 0

    print("All tests passed!")
